### ISSUE TITLE
Change Dart bindings to use callbacks instead of Futures.

### ISSUE THREAD

**Issue by zanderso** [domokit/mojo#777]
That is, the Mojo interface

```mojo
interface Foo {
  Bar() => (int32 a, int32 b, int32 c)
};
```

becomes the Dart interface:

```dart
abstract class Foo {
  void bar(void callback(int a, int b, int c));
}
```

On the proxy side, the callback is invoked when a reply is received. On the Stub side, the callback makes a Response object and sends it.

This is related to the discussion on issue #647. @j9brown @abarth

**Comment by ianloic** [comment]
Quoting @abarth on #647:
>I'd lean towards actually make them have the same interface, probably by dropping Futures from the proxy versions of the interface. In practice, the generated interfaces aren't idiomatic enough to use directly in Dart code anyway because, in mojom, messages can return multiple results, which doesn't map well to Futures, which can resolve only to a single value. The struct the generated bindings introduce to paper over this issue ends up not being something we want to expose to end-developers. If we used callbacks instead of Futures, we could handle multiple results in the same way they're handled in other languages (as multiple parameters to the callback) and the stub and proxy side of the connection could actually have the same interface (again, as in other languages).

Could we just use a Future for a Dart class not a mojom.Struct? Futures feel a lot more idiomatic than callbacks in Dart.

**Comment by j9brown** [comment]
I think a Future-based interface could work nicely if the interface was
designed asymmetrically.

Consider the following Mojom interface:

interface Foo {
   A(int arg1, int arg2);
   B(int arg1, int arg2) => (int result1, int result2);
}

One gotcha with this interface is that it contains methods which have
results and others which don't.  When designing the Dart interfaces, we
need to preserve message ordering.  That means we can't design an interface
where some methods are "async" and others aren't since the "async" method
calls always get deferred to the next event loop.

So this is right out.  (Please forgive me if I get the Dart syntax wrong.)

abstract class BadFoo {
    void A(int arg1, int arg2);
    BResult B(int arg1, int arg2) async;
};

Although convenient for implementors, BadFoo would have problems because A
and B messages could get reordered since A messages would run synchronously
whereas B messages would be deferred.  We could remedy this by making all
methods async but that's kind of annoying.

Instead, we can define asymmetric interfaces like so:

abstract class FooStub {
    void A(int arg1, int arg2);
    void B(int arg1, int arg2, Completer<BResult> completer);
};

abstract class FooProxy {
    void A(int arg1, int arg2);
    Future<BResult> B(int arg1, int arg2);  /* no async keyword */
};

class BResult {
    int result1;
    int result2;
   BResult(int this.result1, int this.result2);
};

All we've really done here is replace the callbacks with more familiar Dart
types, namely a Completer on the implementor's side (stub) and a Future on
the caller's side (proxy).  Slightly annoying for implementors but friendly
and familiar to callers.

With that in mind, we could perhaps ask for language improvements to Dart
to provide syntax sugar for methods which accept Completers as an
argument... ;)

Jeff.

**Comment by abarth** [comment]
@j9brown That's what we've tried up to now, and it doesn't end up working well when you actually try to use the interfaces in Dart.  The problem is `BResult` is unnatural.  We've ended up writing wrappers for the mojom interfaces we use in Dart anyway to make them idiomatic.

**Comment by abarth** [comment]
@ianloic That's what the current code generator does and it ends up being unnatural because the mojom is designed such that replies contain multiple values but `Future`s are designed in Dart to be single-valued.  There are several choices for how to resolve this mismatch:

1. Change mojom to make replies single-valued.  We could then use `Future`s to represent replies without a mismatch, but that would make mojom less expressive and make the interfaces less idiomatic in Go.
2. We could optimize the single-valued reply case to use `Future`s and use some other mechanism (e.g., callbacks) for the multi-valued reply case.  The non-uniformity would mean that interfaces designed with Dart in mind would be more likely to have single-valued replies, which would compromise the language neutrality of mojom.
3. We could use a language construct in Dart that can handle multiple return values (e.g., callbacks).

Of these choices, I think that (3) is least bad.  It also has the advantage that using callbacks makes the proxy and stub interfaces identical in Dart, similar to how they're identical in other languages (e.g., C++ and Java, which also model replies using callbacks).

**Comment by abarth** [comment]
@zanderso That looks about right.  Ideally we'd declare the type for the callback inline, as you have in the example, to avoid polluting the namespace with a (meaningless) typedef for the callback.

**Comment by zanderso** [comment]
We could also generate something equivalent to:

```dart
class FutureyFooProxy {
  FooProxy _proxy;
  FutureyFooProxy(this._proxy);
  Future bar() {
    var completer = new Completer();
    _proxy.bar((a, b, c) {
      completer.complete({'a': a, 'b': b, 'c': c})
    });
    return completer.future;
  }
}
```

For folks who like Futures, and don't mind Maps and the extra trip through the event loop. For calls that return a single result, we could give a more specific return type and skip the map.

@tonygentilcore, @jimbeveridge, @alhaad, @HansMuller  it would be great to hear your thoughts on this as well. How difficult would it be to port your existing code to this?

**Comment by j9brown** [comment]
Why do we need maps here?

**Comment by zanderso** [comment]
@j9brown It's not needed, I was just jotting down the idea quickly.

**Comment by abarth** [comment]
@zanderso Maps are worse than structs for multiplexing multiple values into single-valued Futures.  For starters, they're less efficient and not type-safe.

**Comment by zanderso** [comment]
Update example to clarify.

**Comment by floitschG** [comment]
@j9brown "async" is an implementation detail (which is why it is after the signature of the function). It is syntactic sugar for the body. This also means that you can't return `BResult` from a function that is implemented with `async` since it will automatically return a `Future`.
A function that returns a `Future` will always return the future immediately, and will always require at least one microtask tick before the future is completed. A function that uses the `async` feature is furthermore delayed by one microtask tick before it even starts executing its body.

Wrt the `FutureyFooProxy`: this would be an additional class. We are basically exploring the idea of generating a simple useful class for the common case, while still providing a lower level, less darty, solution for the more complicated case.

**Comment by krisgiesing** [comment]
So basically the idea is to provide both a callback-based interface and a Future-based interface, and clients decide which style they like better?

My first reaction is that it could be more confusing than it's worth, given that it's already a bit difficult to find good descriptive names for the generated APIs. But at the very least, we could document this pattern so that developers who want the Future style can easily create one for the particular API they care about.

**Comment by pylaligand** [comment]
Should we choose to switch to callbacks or use the futurey thing, I don't think it would be a big change for us.

**Comment by abarth** [comment]
> So basically the idea is to provide both a callback-based interface and a Future-based interface, and clients decide which style they like better?

IMHO, the generated code should be opinionated about which of these styles to generate.  Specifically, the generated code should contain only one of them.

>  But at the very least, we could document this pattern so that developers who want the Future style can easily create one for the particular API they care about

+1 to documenting this pattern rather than generate code for it.

**Comment by j9brown** [comment]
Right.  I figured the async method syntax was sugar.  What I was getting at
is that we probably don't want clients to implement an interface that
returns Futures only for methods which have results.  It would be pretty
tempting to write code that behaves unexpectedly by mixing async and
non-async methods in the same interface implementation.

That's why I opted for the Completer on the stub interface.  Not sure
that's really the best thing though.  I don't know whether Dart programmers
would typically trip over this though.  :)
